### ISSUE TITLE
Negative Entropy in TRL PPOv2Trainer TLDR Example

### ISSUE THREAD

**Issue by RylanSchaeffer** [huggingface/trl#2022]
### System Info

- `transformers` version: 4.44.0
- Platform: Linux-5.4.0-162-generic-x86_64-with-glibc2.31
- Python version: 3.11.9
- Huggingface_hub version: 0.23.4
- Safetensors version: 0.4.3
- Accelerate version: 0.32.1
- Accelerate config:    - compute_environment: LOCAL_MACHINE
        - distributed_type: FSDP
        - mixed_precision: bf16
        - use_cpu: False
        - debug: True
        - num_processes: 2
        - machine_rank: 0
        - num_machines: 1
        - rdzv_backend: static
        - same_network: True
        - main_training_function: main
        - enable_cpu_affinity: False
        - fsdp_config: {'fsdp_activation_checkpointing': True, 'fsdp_auto_wrap_policy': 'TRANSFORMER_BASED_WRAP', 'fsdp_backward_prefetch': 'BACKWARD_PRE', 'fsdp_cpu_ram_efficient_loading': True, 'fsdp_forward_prefetch': True, 'fsdp_offload_params': True, 'fsdp_sharding_strategy': 'FULL_SHARD', 'fsdp_state_dict_type': 'SHARDED_STATE_DICT', 'fsdp_sync_module_states': True, 'fsdp_use_orig_params': True}
        - downcast_bf16: no
        - tpu_use_cluster: False
        - tpu_use_sudo: False
        - tpu_env: []
        - dynamo_config: {'dynamo_backend': 'EAGER'}
- PyTorch version (GPU?): 2.4.0+cu121 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using distributed or parallel set-up in script?: Yes
- Using GPU in script? Yes
- GPU type: NVIDIA A100-SXM4-80GB

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder
- [ ] My own task or dataset (give details below)

### Reproduction

In [TRL's PPOv2Trainer TLDR example](https://github.com/huggingface/trl/blob/main/examples/scripts/ppo/ppo_tldr.py), run the default command:

https://github.com/huggingface/trl/blob/main/examples/scripts/ppo/ppo_tldr.py#L32-L44 

### Expected behavior

Entropy must be non-negative. However, when I run the official example, the entropy is negative:

![image](https://github.com/user-attachments/assets/60aef46e-bc2d-4842-a69c-99a6828f8a8e)

I don't think I'm making a mistake because this negative entropy also appears in the official documentation at https://huggingface.co/docs/trl/ppov2_trainer. Specifically, look early in training, at maybe 20k episodes:

![image](https://github.com/user-attachments/assets/cfb361bc-ed62-4e0a-b484-a59c1b1e92fc)

The [documentation](https://huggingface.co/docs/trl/ppov2_trainer) describes `objective/entropy` as "The mean entropy of the policy, indicating the randomness of the actions chosen by the policy." If this is incorrect, and some other quantity is computed instead, then perhaps the documentation needs to be updated?

**Comment by RylanSchaeffer** [comment]
I don't know if this is the culprit, but I noticed that I'm using `bf16` as does the tutorial, and in `bf16`, the two following quantities don't agree:

`torch.einsum("bse,bse->bs", prob_dist, logits)`

`torch.sum(prob_dist * logits, dim=-1)`

**Comment by RylanSchaeffer** [comment]
Following [this previous PR](https://github.com/huggingface/trl/pull/156), it might be worthwhile to consider upcasting the tensors before computing logged quantities.

But I don't know if this explains how the entropy is becoming negative...

**Comment by RylanSchaeffer** [comment]
On another PPOv2 run, I again observe negative entropy: 

![image](https://github.com/user-attachments/assets/7e2f447f-697b-465c-a4e9-603b4d0842ae)

**Comment by lwaekfjlk** [comment]
same question here.

**Comment by qgallouedec** [comment]
Hey @RylanSchaeffer, sorry for the very late reply, catching up on the backlog.

You're right that there's a real bug here, but the root cause isn't bf16 precision. `objective/entropy` is computed as `(-logprobs).sum(1).mean()` after the padding positions in `logprobs` have been masked to the sentinel `INVALID_LOGPROB = 1.0`  and so every padding step contributes `-1.0` to the per-sequence sum and drags the mean negative as soon as responses are short relative to `response_length`, which is exactly what TLDR rollouts look like. The `policy/entropy_avg` metric (computed from the logits) is unaffected. This is the same root cause as #2281; the fix that landed in #3801 only covered the RLOO rewrite, not PPO, note that PPO has since been moved to `trl.experimental.ppo` but the offending line at https://github.com/huggingface/trl/blob/51c6d3ca31fb4cc80ff719c0844bbdfcd0feeefb/trl/experimental/ppo/ppo_trainer.py#L891 is unchanged. This issue is open for contributions.
